kythe/go/platform/kcd/kzipdb/kzipdb_test.go

```diff
@@ -0,0 +1,198 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package kzipdb
+
+import (
+	"bytes"
+	"context"
+	"sort"
+	"strings"
+	"sync"
+	"testing"
+
+	"kythe.io/kythe/go/platform/kcd"
+	"kythe.io/kythe/go/platform/kcd/kythe"
+	"kythe.io/kythe/go/platform/kzip"
+
+	"github.com/google/go-cmp/cmp"
+
+	apb "kythe.io/kythe/proto/analysis_go_proto"
+	spb "kythe.io/kythe/proto/storage_go_proto"
+)
+
+var testInput struct {
+	sync.Once
+
+	data  []byte
+	units []string
+	files []string
+}
+
+func newUnit(key, lang, corpus string) *apb.CompilationUnit {
+	return &apb.CompilationUnit{
+		VName: &spb.VName{
+			Corpus:   corpus,
+			Language: lang,
+		},
+		OutputKey: key,
+	}
+}
+
+func newIndex(revs ...string) *apb.IndexedCompilation_Index {
+	return &apb.IndexedCompilation_Index{Revisions: revs}
+}
+
+func setup(t *testing.T) {
+	t.Helper()
+	testInput.Do(func() {
+
+		buf := bytes.NewBuffer(nil)
+		w, err := kzip.NewWriter(buf)
+		if err != nil {
+			t.Fatal("NewWriter: %v", err)
+		}
+		for _, ic := range []*apb.IndexedCompilation{
+			{Unit: newUnit("A", "go", "kythe"), Index: newIndex("123")},
+			{Unit: newUnit("B", "c++", "kythe"), Index: newIndex("456", "789")},
+			{Unit: newUnit("C", "go", "chromium"), Index: newIndex("123")},
+			{Unit: newUnit("D", "protobuf", "kythe"), Index: newIndex("789")},
+		} {
+			digest, err := w.AddUnit(ic.Unit, ic.Index)
+			if err != nil {
+				t.Fatalf("AddUnit %v: %v", ic.Unit, err)
+			}
+			testInput.units = append(testInput.units, digest)
+		}
+		for _, src := range []string{
+			"apple", "pear", "plum", "cherry",
+		} {
+			digest, err := w.AddFile(strings.NewReader(src))
+			if err != nil {
+				t.Fatalf("AddFile %q: %v", src, err)
+			}
+			testInput.files = append(testInput.files, digest)
+		}
+		if err := w.Close(); err != nil {
+			t.Fatalf("Close: %v", err)
+		}
+
+		testInput.data = buf.Bytes()
+	})
+}
+
+func TestReader(t *testing.T) {
+	setup(t)
+	r, err := kzip.NewReader(bytes.NewReader(testInput.data), int64(len(testInput.data)))
+	if err != nil {
+		t.Fatalf("NewReader failed: %v", err)
+	}
+	db := DB{Reader: r}
+	ctx := context.Background()
+
+	t.Run("NoMatchingRevs", func(t *testing.T) {
+		db.Revisions(ctx, &kcd.RevisionsFilter{Corpus: "bogus"}, func(rev kcd.Revision) error {
+			t.Errorf("Unexpected revision found: %v", rev)
+			return nil
+		})
+	})
+	t.Run("MatchSomeRevs", func(t *testing.T) {
+		want := []kcd.Revision{
+			{Revision: "123", Corpus: "kythe"},
+			{Revision: "456", Corpus: "kythe"},
+			{Revision: "789", Corpus: "kythe"},
```

Comment:
Can you add a revision that doesn't have a kythe corpus so at least one revision doesn't match here?

Reply:
These are the expected values, not the input values—the input values above already contain some non-matching values.

Reply:
Aren't the inputs values:
{Unit: newUnit("A", "go", "kythe"), Index: newIndex("123")},
+			{Unit: newUnit("B", "c++", "kythe"), Index: newIndex("456", "789")},
+			{Unit: newUnit("C", "go", "chromium"), Index: newIndex("123")},
+			{Unit: newUnit("D", "protobuf", "kythe"), Index: newIndex("789")},

if so, there aren't any revisions that aren't 123, 456, or 789

Reply:
Ah, I see what you mean. I thought you wanted another output value here. 👍

Note that we're comparing "whole revisions", i.e., revision + corpus, so a mismatch should not be possible. Nevertheless, I added an explicit case to guard against this. Thanks!